811 gobierto data 2nd iteration

entantoencuanto opened the pull request:
Closes #N


## :v: What does this PR do?

Second iteration of gobierto data module:
- Updates `GobiertoData::Dataset` model including a `create_dataset_from_file` methods which allows to load data into dataset table from a csv file with the option to use a schema file (documentation pending)
- Adds `GobiertoData::Query` and `Gobierto::Data::Visualization` models, forms, API controller, etc.
- Adds fixtures and tests for datasets, queries and visualizations models, forms and API controllers

## :mag: How should this be manually tested?

Follow the documentation in: https://github.com/PopulateTools/issues/wiki/Gobierto-Data

## :shipit: Does this PR changes any configuration file?

No

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

- [ ] new site configuration variable?
- [ ] new site template?
- [x] new module/submodule settings? Documentation is required for module connection setting and `create_dataset_from_file` Dataset model method usage
- [ ] significant changes in some feature?

entantoencuanto committed:
Add test_connection_config class method to GobiertoData::Connection

entantoencuanto committed:
Add settings form to gobierto data admin

entantoencuanto committed:
Add integration tests for GobiertoData module settings

entantoencuanto committed:
Povide meta data in API data query responses

entantoencuanto committed:
Add data API support to CSV format

entantoencuanto committed:
Add tests for csv format

entantoencuanto committed:
Force inline content disposition of csv data API response

entantoencuanto committed:
Extract csv response details to a method

entantoencuanto committed:
Use string_output instance_option on serializers with custom fields

entantoencuanto committed:
Define csv output of AR relations using serializers with custom fields

entantoencuanto committed:
Define Gobierto data datasets controller and serializers

entantoencuanto committed:
Update tests for value_string method of custom field records

entantoencuanto committed:
Change datasets meta response

entantoencuanto committed:
Generate @resource from custom field api concern if not present

entantoencuanto committed:
Include custom field filters in datasets index

ferblape commented:
Ping me when I can review this.

entantoencuanto committed:
Add function to create table datasets from csv files

entantoencuanto committed:
Use csv_separator param on inspect_csv_schema method

entantoencuanto committed:
Use output type of functions in transformed and destination tables

entantoencuanto committed:
Trim raw values in transformation funcitions by default

entantoencuanto committed:
Add transformation function for number strings with decimal separator other than .

ferblape committed:
Disable CORS in API requests

ferblape commented:
I have added a commit to disable CORS (see https://github.com/PopulateTools/issues/issues/837)

![Screen Shot 2019-12-11 at 11 48 50](https://user-images.githubusercontent.com/17616/70615229-50436200-1c0c-11ea-8e5e-0bc97d116569.png)

Can I deploy this branch to staging?

entantoencuanto committed:
Add query and visualization models

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
Define form for queries and move validations from model

entantoencuanto committed:
Change datasets API to include queries

entantoencuanto committed:
Define query and user serializers

entantoencuanto committed:
Add queries controller

entantoencuanto committed:
Add missing associations

entantoencuanto committed:
Add serializer, form and controller for visualizations API

entantoencuanto committed:
Remove custom fields support from queries

entantoencuanto committed:
Take user and dataset from query in form if ids are not provided

entantoencuanto committed:
Use params provided in the body on queries controller

Instead of using dataset_slug param provided by the path use a
dataset_id present in the body json on creation and ignore both user_id
and dataset_id on update.
The requests also accepts a dataset_id query param to filter queries by
dataset.

entantoencuanto committed:
Take user and query from visualization in form if ids are not provided

entantoencuanto committed:
Use params provided in the body on visualizations controller

Instead of using dataset_slug and query_id params provided by the path
use a query_id present in the body json on creation and ignore both
user_id and query_id on update.
The requests also accepts a query_id and a dataset_id query params to
filter visualizations by query or dataset.

entantoencuanto committed:
Un-nest queries and visualizations routes

entantoencuanto committed:
Update links of queries and visualizations

entantoencuanto committed:
Update comments with paths

entantoencuanto committed:
Add some links to API responses of GobiertoData

entantoencuanto committed:
Fix relation for visualizations new action

entantoencuanto committed:
Change filter syntax of queries and visualizations and add user_id

entantoencuanto committed:
Improve use of @item in API controllers

entantoencuanto committed:
Accept translatable attributes both as translations hash and as string

The API accepts both name_translations with a translations hash and name
with a string as value. If provided, name_translations is used to
create or update the resource and name is ignored. When name is provided
the value is used for translation of current locale

entantoencuanto committed:
Add fixtures for datasets, queries and visualizations

entantoencuanto committed:
Add basic tests for gobierto data models

entantoencuanto committed:
Avoid exceptions in forms when site is not present

entantoencuanto committed:
Add query and visualization form tests

entantoencuanto committed:
Allow to pass current_site to DatasetSerializer from instance_options and use model instead of relation

entantoencuanto committed:
Add dataset fixture

entantoencuanto committed:
Add visualizations association to DatasetMetaSerializer

entantoencuanto committed:
Add datasets API tests

entantoencuanto committed:
Add exclude_relationships option in serializers for csv export

entantoencuanto committed:
Change filter_params in visualizations controller to filter by dataset

entantoencuanto committed:
Add tests for queries and visualizations gobierto data API

entantoencuanto commented:
@ferblape, I think it's ready for review

ferblape commented:
Perfect!

El vie., 20 dic. 2019 14:24, Eduardo Martínez <notifications@github.com>
escribió:

> @ferblape <https://github.com/ferblape>, I think it's ready to merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PopulateTools/gobierto/pull/2691?email_source=notifications&email_token=AAAEJUCX7VNN4TT7ZWBNWL3QZTBRVA5CNFSM4JQOITZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHM5D2A#issuecomment-567923176>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEJUBZJHHBTLJNW7GKRT3QZTBRVANCNFSM4JQOITZA>
> .
>

ferblape commented on app/models/gobierto_data/dataset.rb:
```diff
@@ -32,6 +34,22 @@ def rails_model
                        end
     end
 
+    def create_dataset_from_file(file_path, schema_file: nil, export_files: false, csv_separator: ",")
```
I think this name is confusing, because it doesn't create a Dataset instance, but loads a dataset into the database

ferblape commented on config/routes.rb:
```diff
@@ -561,7 +565,21 @@
         namespace :api, path: "/" do
           namespace :v1, constraints: ::ApiConstraint.new(version: 1, default: true), path: "/api/v1/data" do
-            get "/" => "query#index", as: :root
+            get "/" => "query#index", as: :root, defaults: { format: "json" }
```
This route, according to the docs should be `
GET /api/v1/data/data{.json|.csv)`

ferblape commented:
Just review my comments and feel free to merge.

entantoencuanto committed:
Fix gobierto_data root query path

entantoencuanto committed:
Rename method to import data into a dataset from file
